### PR TITLE
feat: additional device info

### PR DESCRIPTION
**Pull request by anujagrawal06**

### CODE REVIEW

File: TrackierSDK/example-app-kotlin/app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -2,6 +2,7 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     package="com.trackier.example_app_kotlin">
 
+    <uses-permission android:name="android.permission.GET_ACCOUNTS" />
```
Comment (Hemant-Mann) [inline TrackierSDK/example-app-kotlin/app/src/main/AndroidManifest.xml:5]: let's remove this permission @anujagrawal06

File: TrackierSDK/example-app-kotlin/app/src/main/java/com/trackier/example_app_kotlin/MainActivity.kt [hunk 2]
```diff
@@ -57,13 +76,239 @@ class MainActivity : AppCompatActivity(){
...
     }
+
+    @SuppressLint("QueryPermissionsNeeded")
+    fun getDeviceInstallApplication(): String {
+        val pm = packageManager
+        val apps = pm.getInstalledApplications(PackageManager.GET_META_DATA)
+        return apps.joinToString()
+    }
+
+    private fun checkIsEmulator(): Boolean {
```
Comment (Hemant-Mann) [inline TrackierSDK/example-app-kotlin/app/src/main/java/com/trackier/example_app_kotlin/MainActivity.kt:94]: remove additional functions from this class

File: TrackierSDK/example-app-kotlin/app/src/main/java/com/trackier/example_app_kotlin/MainActivity.kt [hunk 3]
```diff
@@ -57,13 +76,239 @@ class MainActivity : AppCompatActivity(){
...
+                Log.d("TAG", "onCreate: " + getDeepLinkParams(uri).toString())
             }
         }
 
         //requestPermission()
 
     }
+
+    @SuppressLint("QueryPermissionsNeeded")
+    fun getDeviceInstallApplication(): String {
```
Comment (Hemant-Mann) [inline TrackierSDK/example-app-kotlin/app/src/main/java/com/trackier/example_app_kotlin/MainActivity.kt:88]: remove

File: TrackierSDK/example-app-kotlin/app/src/main/java/com/trackier/example_app_kotlin/MainActivity.kt [hunk 4]
```diff
@@ -57,13 +76,239 @@ class MainActivity : AppCompatActivity(){
...
+                || Build.PRODUCT.contains("sdk")
+                || Build.PRODUCT.contains("sdk_x86")
+                || Build.PRODUCT.contains("sdk_gphone64_arm64")
+                || Build.PRODUCT.contains("vbox86p")
+                || Build.PRODUCT.contains("emulator")
+                || Build.PRODUCT.contains("simulator"))
+    }
+
+
+    fun getIpv4HostAddress(): String {
```
Comment (Hemant-Mann) [inline TrackierSDK/example-app-kotlin/app/src/main/java/com/trackier/example_app_kotlin/MainActivity.kt:115]: remove

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt [hunk 5]
```diff
@@ -80,6 +108,7 @@ data class DeviceInfo(
             deviceInfo.screenSize = screenSize(screenLayout)
             deviceInfo.screenFormat = screenFormat(screenLayout)
             deviceInfo.screenDensity = screenDensity(displayMetrics.densityDpi)
+            // deviceInfo.screenDensityNumber = displayMetrics.densityDpi
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt:111]: remove

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt [hunk 6]
```diff
@@ -88,8 +117,152 @@ data class DeviceInfo(
...
+            deviceInfo.systemVolume = getSystemVolume(context)
+            deviceInfo.orientation = getDeviceOrientation(context)
+            deviceInfo.bootTime = getDeviceBootTime();
+            if (VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
+                deviceInfo.availableInternalStorage = getAvailableInternalStorage()
+                deviceInfo.totalInternalStorage = getTotalInternalStorage()
+            }
+            deviceInfo.cpuDetail = getCPUDetails()
+            deviceInfo.chargingStatus = getDeviceChargingStatus(context)
+            if (VERSION.SDK_INT >= Build.VERSION_CODES.M)
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt:135]: missing parenthesis for if statement

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt [hunk 7]
```diff
@@ -88,8 +117,152 @@ data class DeviceInfo(
...
+        fun getIpv4HostAddress(): String {
+            NetworkInterface.getNetworkInterfaces()?.toList()?.map { networkInterface ->
+                networkInterface.inetAddresses?.toList()?.find {
+                    !it.isLoopbackAddress && it is Inet4Address
+                }?.let { return it.hostAddress }
+            }
+            return ""
+        }
+
+        var strOrientation: String? = null
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt:160]: var outside the function?

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt [hunk 8]
```diff
@@ -88,8 +117,152 @@ data class DeviceInfo(
...
+        fun getDeviceOrientation(context: Context): String {
+            var orientation = context.resources.configuration.orientation
+            if (orientation == Configuration.ORIENTATION_PORTRAIT)
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt:163]: plz use proper parenthesis for if else

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt [hunk 9]
```diff
@@ -88,8 +117,152 @@ data class DeviceInfo(
...
+            val iStat = StatFs(iPath.path)
+            val iBlockSize = iStat.blockSizeLong
+            val iTotalBlocks = iStat.blockCountLong
+            val iTotalSpace = iTotalBlocks * iBlockSize
+            return "$iTotalSpace"
+
+        }
+
+        fun getTotalMemory(context: Context): String {
+            val actManager = context.getSystemService(ACTIVITY_SERVICE) as ActivityManager
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt:181]: we are doing a lot of things duplicate, merge the two functions into one

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt [hunk 10]
```diff
@@ -88,8 +117,152 @@ data class DeviceInfo(
...
+            val iStat = StatFs(iPath.path)
+            val iBlockSize = iStat.blockSizeLong
+            val iAvailableBlocks = iStat.availableBlocksLong
+            val iAvailableSpace = iAvailableBlocks * iBlockSize
+            //   val iAvailableSpace = formatSize(iAvailableBlocks * iBlockSize)
+            return "$iAvailableSpace"
+
+        }
+
+        private fun formatSize(size: Long): String? {
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt:233]: remove this function as it is not used anywhere

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt [hunk 11]
```diff
@@ -88,8 +117,152 @@ data class DeviceInfo(
...
+                resultBuffer.insert(commaOffset, ',')
+                commaOffset -= 3
+            }
+            if (suffix != null) resultBuffer.append(suffix)
+            return "$resultBuffer"
+        }
+
+        @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
+        fun getBatteryLevel(context: Context): String {
+            val bm = context.getSystemService(BATTERY_SERVICE) as BatteryManager
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt:256]: wrap these code in try catch blocks @anujagrawal06

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt [hunk 12]
```diff
@@ -139,6 +312,60 @@ data class DeviceInfo(
             }
         }
 
+        @RequiresApi(Build.VERSION_CODES.M)
+        private fun getHeadphonesPlugged(context: Context): String {
+            val audioManager = context.getSystemService(AUDIO_SERVICE) as AudioManager
+            val audioDevices = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+                audioManager.getDevices(AudioManager.GET_DEVICES_INPUTS)
+            } else {
+                TODO("VERSION.SDK_INT < M")
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt:321]: remove todo

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt [hunk 13]
```diff
@@ -139,6 +312,60 @@ data class DeviceInfo(
...
+                    cpuDetails += String(bArray) //Stroing all the details in cpuDetails
+                }
+                `is`.close()
+            } catch (ex: IOException) {
+                ex.printStackTrace()
+            }
+            return cpuDetails
+        }
+
+        private fun screenDensityNumer(density: Int): String? {
```
Comment (Hemant-Mann) [inline TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt:358]: return this number density

File: TrackierSDK/example-app-kotlin/app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (anujagrawal06) [reply TrackierSDK/example-app-kotlin/app/src/main/AndroidManifest.xml]: removed getAccount permission

File: TrackierSDK/example-app-kotlin/app/src/main/java/com/trackier/example_app_kotlin/MainActivity.kt (hunk 2 above)
Comment (anujagrawal06) [reply TrackierSDK/example-app-kotlin/app/src/main/java/com/trackier/example_app_kotlin/MainActivity.kt]: removed

File: TrackierSDK/example-app-kotlin/app/src/main/java/com/trackier/example_app_kotlin/MainActivity.kt (hunk 3 above)
Comment (anujagrawal06) [reply TrackierSDK/example-app-kotlin/app/src/main/java/com/trackier/example_app_kotlin/MainActivity.kt]: removed

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt (hunk 5 above)
Comment (anujagrawal06) [reply TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt]: removed

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt (hunk 6 above)
Comment (anujagrawal06) [reply TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt]: Done

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt (hunk 7 above)
Comment (anujagrawal06) [reply TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt]: Done

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt (hunk 8 above)
Comment (anujagrawal06) [reply TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt]: Done

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt (hunk 10 above)
Comment (anujagrawal06) [reply TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt]: removed

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt (hunk 9 above)
Comment (anujagrawal06) [reply TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt:185]: Done

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt (hunk 11 above)
Comment (anujagrawal06) [reply TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt]: Done

File: TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt (hunk 12 above)
Comment (anujagrawal06) [reply TrackierSDK/sdk-core/src/main/java/com/trackier/sdk/DeviceInfo.kt]: Done

**Review by Hemant-Mann — APPROVED** [review]

**Pull request merged**